### PR TITLE
thread safety added to schedule_work and work_item_done APIs

### PR DESCRIPTION
**Pull request by sarab96**
Both APIs schedule_work and work_item_done are allowed to be used asynchronously in different threads.
So added thread safety.

@sonals Please check
@maxzhen Please do the needful

### CODE REVIEW

File: src/xma/include/lib/xmahw.h [hunk 1]
```diff
@@ -47,12 +49,23 @@ typedef struct XmaHwKernel
     uint32_t    ddr_bank;
     //For execbo:
     int32_t     kernel_complete_count;
+    std::unique_ptr<std::condition_variable> kernel_complete_count_cv;
```
Comment (stsoe) [inline src/xma/include/lib/xmahw.h]: Why pointer?  Also, I don't see the condition variable used any where?

File: src/xma/include/lib/xmahw.h [hunk 2]
```diff
@@ -47,12 +49,23 @@ typedef struct XmaHwKernel
...
+    std::unique_ptr<std::atomic<bool>> kernel_complete_count_locked;
+
     void*       kernel_cmd_queue;
     void*       kernel_cmd_completion_queue;
     uint32_t    kernel_execbo_handle[MAX_EXECBO_POOL_SIZE];
     char*       kernel_execbo_data[MAX_EXECBO_POOL_SIZE];
     bool        kernel_execbo_inuse[MAX_EXECBO_POOL_SIZE];
     uint32_t    reserved[16];
+
+    XmaHwKernel(): kernel_complete_count_cv(new std::condition_variable),
```
Comment (stsoe) [inline src/xma/include/lib/xmahw.h]: Is there any reason why this is heap allocated?

File: src/xma/include/lib/xmahw.h [hunk 3]
```diff
@@ -47,12 +49,23 @@ typedef struct XmaHwKernel
...
+      kernel_complete_count_locked(new std::atomic<bool>) {
```
Comment (stsoe) [inline src/xma/include/lib/xmahw.h]: heap allocation?

**Comment by maxzhen** [comment]
ok to test

File: src/xma/include/lib/xmahw.h (hunk 1 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: I will remove unused condition_variable.
I'm changing c code to c++ in small portions. Later I may change arrays to vector so I want to keep it as pointer.

File: src/xma/include/lib/xmahw.h (hunk 2 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: I will remove condition_variable

File: src/xma/include/lib/xmahw.h (hunk 3 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: I think it is okay to have it on heap. Later I may change arrays to vector. So it should be fine keeping in mind overall code architecture. Any suggestions?

Comment (stsoe) [reply src/xma/include/lib/xmahw.h]: Yes, please don't use heap allocation unless you need to. Array or not it doesn't matter; at least I can't see why.

File: src/xma/include/lib/xmahw.h (hunk 1 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: changed

File: src/xma/include/lib/xmahw.h (hunk 2 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: changed

File: src/xma/include/lib/xmahw.h (hunk 3 above)
Comment (sarab96) [reply src/xma/include/lib/xmahw.h]: I looked at the code. Later on vector will be used to hold this struct. Num of structs are not fixed as it depends on xclbin. There will be one struct for one CU. So number of std::atomic depends on number of CUs in an xclbin.
Though I have changed to make_unique for better protection. I think this is fine now. Please check.

**Comment by sarab96** [comment]
retest this please

**Comment by sarab96** [comment]
Will resubmit after changing HwSession to internal private structure

**Pull request closed without merging**